drivers/frequency/adf4371/adf4371.c

```diff
@@ -0,0 +1,820 @@
+
+	/* Calculate Timeouts */
+	vco_band_div = DIV_ROUND_UP(dev->fpfd, 2400000U);
+
+	tmp = DIV_ROUND_CLOSEST(dev->fpfd, 1000000U);
+	do {
+		timeout++;
+		if (timeout > 1023) {
+			timeout = 2;
+			synth_timeout++;
+		}
+	} while (synth_timeout * 1024 + timeout <= 20 * tmp);
+
+	do {
+		vco_alc_timeout++;
+	} while (vco_alc_timeout * 1024 - timeout <= 50 * tmp);
+
+	dev->buf[0] = vco_band_div;
+	dev->buf[1] = timeout & 0xFF;
+	dev->buf[2] = ADF4371_TIMEOUT(timeout >> 8) | 0x04;
+	dev->buf[3] = synth_timeout;
+	dev->buf[4] = ADF4371_VCO_ALC_TOUT(vco_alc_timeout);
+
+	ret = adf4371_write_bulk(dev, ADF4371_REG(0x30), dev->buf, 5);
+	if (ret < 0)
+		return 0;
+
+	return adf4371_channel_config(dev);
+}
+
+/**
+ * Recalculate rate corresponding to a channel.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_recalc_rate(struct adf4371_dev *dev, uint32_t chan,
+				uint64_t *rate)
+{
+	if (chan > dev->num_channels)
+		return FAILURE;
+
+	*rate = adf4371_pll_fract_n_get_rate(dev, chan);
+
+	return SUCCESS;
+}
+
+/**
+ * Calculate closest possible rate
+ * @param dev - The device structure
+ * @param rate - The desired rate.
+ * @param rounded_rate - The closest possible rate of desired rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_round_rate(struct adf4371_dev *dev, uint64_t rate,
+			       uint64_t *rounded_rate)
+{
+	*rounded_rate = rate;
+
+	return SUCCESS;
+}
+
+/**
+ * Set channel rate.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_set_rate(struct adf4371_dev *dev, uint32_t chan,
+			     uint64_t rate)
+{
+	if (chan >= dev->num_channels)
+		return FAILURE;
+
+	return adf4371_set_freq(dev, rate, chan);
+}
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_init(struct adf4371_dev **device,
+		     const struct adf4371_init_param *init_param)
+{
+	struct adf4371_dev *dev;
+	int32_t ret;
+	uint32_t i;
+
+	dev = (struct adf4371_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret < 0)
+		return ret;
+
+	dev->spi_3wire_en = init_param->spi_3wire_enable;
+
+	dev->clkin_freq = init_param->clkin_frequency;
+
+	dev->differential_ref_clk = init_param->differential_ref_clock;
+
+	if (init_param->loop_filter_inverting)
+		dev->pd_pol = 0;
+	else
+		dev->pd_pol = 1;
+
+	if (init_param->charge_pump_microamp == 0) {
+		/* Default charge pump current is 1.75mA */
+		dev->cp_settings.icp = 1750;
```

Comment:
these values look a bit cryptic to me, maybe comment them more to see what they correspond to in datasheet or defined them as macros